plugins/experimental/statichit/statichit.cc

```diff
@@ -0,0 +1,637 @@
+    TSMBufferDestroy(this->buffer);
+  }
+};
+
+struct StaticHitRequest {
+  StaticHitRequest() {}
+
+  off_t nbytes        = 0; // Number of bytes to generate.
+  unsigned maxAge     = 0; // Max age for cache responses.
+  unsigned statusCode = 200;
+  IOChannel readio;
+  IOChannel writeio;
+  StaticHitHttpHeader rqheader;
+
+  std::string body;
+  std::string mimeType;
+
+  static StaticHitRequest *
+  createStaticHitRequest(StaticHitConfig *tc)
+  {
+    StaticHitRequest *shr = new StaticHitRequest;
+    std::ifstream ifstr;
+
+    ifstr.open(tc->filePath);
+    if (!ifstr) {
+      shr->statusCode = tc->failureCode;
+      return shr;
+    }
+
+    std::stringstream sstr;
+    sstr << ifstr.rdbuf();
+    shr->body       = sstr.str();
+    shr->nbytes     = shr->body.size();
+    shr->mimeType   = tc->mimeType;
+    shr->statusCode = tc->successCode;
+    shr->maxAge     = tc->maxAge;
+
+    return shr;
+  }
+
+  ~StaticHitRequest() = default;
+};
+
+// Destroy a StaticHitRequest, including the per-txn continuation.
+static void
+StaticHitRequestDestroy(StaticHitRequest *trq, TSVIO vio, TSCont contp)
+{
+  if (vio) {
+    TSVConnClose(TSVIOVConnGet(vio));
+  }
+
+  TSContDestroy(contp);
+  delete trq;
+}
+
+// NOTE: This will always append a new "field_name: value"
+static void
+HeaderFieldDateSet(const StaticHitHttpHeader &http, const char *field_name, int64_t field_len, time_t value)
+{
+  TSMLoc field;
+
+  TSMimeHdrFieldCreateNamed(http.buffer, http.header, field_name, field_len, &field);
+  TSMimeHdrFieldValueDateSet(http.buffer, http.header, field, value);
+  TSMimeHdrFieldAppend(http.buffer, http.header, field);
+  TSHandleMLocRelease(http.buffer, http.header, field);
+}
+
+// NOTE: This will always append a new "field_name: value"
+static void
+HeaderFieldIntSet(const StaticHitHttpHeader &http, const char *field_name, int64_t field_len, int64_t value)
+{
+  TSMLoc field;
+
+  TSMimeHdrFieldCreateNamed(http.buffer, http.header, field_name, field_len, &field);
+  TSMimeHdrFieldValueInt64Set(http.buffer, http.header, field, -1, value);
+  TSMimeHdrFieldAppend(http.buffer, http.header, field);
+  TSHandleMLocRelease(http.buffer, http.header, field);
+}
+
+// NOTE: This will always append a new "field_name: value"
+static void
+HeaderFieldStringSet(const StaticHitHttpHeader &http, const char *field_name, int64_t field_len, const char *value)
+{
+  TSMLoc field;
+
+  TSMimeHdrFieldCreateNamed(http.buffer, http.header, field_name, field_len, &field);
+  TSMimeHdrFieldValueStringSet(http.buffer, http.header, field, -1, value, -1);
+  TSMimeHdrFieldAppend(http.buffer, http.header, field);
+  TSHandleMLocRelease(http.buffer, http.header, field);
+}
+
+static void
+WriteResponseHeader(StaticHitRequest *trq, TSCont contp, TSHttpStatus status)
+{
+  StaticHitHttpHeader response;
+
+  VDEBUG("writing response header");
+
+  TSReleaseAssert(TSHttpHdrTypeSet(response.buffer, response.header, TS_HTTP_TYPE_RESPONSE) == TS_SUCCESS);
```

Comment:
I think we frown on putting code with required side effects in asserts. This has been a serious problem in the core.

Reply:
This is a common idiom in our plugins and examples.

Reply:
Yes, but I think it shouldn't be.